File: src/main/osu/beatmap/parser/sections/BeatmapGeneralParser.java
package main.osu.beatmap.parser.sections;

import com.rian.difficultycalculator.beatmap.constants.SampleBank;

import main.osu.Utils;
import main.osu.beatmap.BeatmapData;
import main.osu.beatmap.constants.BeatmapCountdown;

/**
 * A parser for parsing a beatmap's general section.
 */
public class BeatmapGeneralParser extends BeatmapKeyValueSectionParser {
    @Override
    public boolean parse(BeatmapData data, String line) {
        String[] p = splitProperty(line);

        switch (p[0]) {
            case "AudioFilename":
                data.general.audioFilename = p[1];
                break;
            case "AudioLeadIn":
                data.general.audioLeadIn = Utils.tryParseInt(p[1], data.general.audioLeadIn);
                break;
            case "PreviewTime":
                data.general.previewTime = Utils.tryParseInt(p[1], data.general.previewTime);
                break;
            case "Countdown":
                data.general.countdown = BeatmapCountdown.parse(p[1]);
                break;
            case "SampleSet":
                data.general.sampleBank = SampleBank.parse(p[1]);
                break;
            case "StackLeniency":
                data.general.stackLeniency = Utils.tryParseFloat(p[1], data.general.stackLeniency);
                break;
        }

        return true;
    }
}

File: src/com/osudroid/beatmaps/parser/sections/BeatmapGeneralParser.kt
package com.osudroid.beatmaps.parser.sections

import com.osudroid.beatmaps.Beatmap
import com.osudroid.beatmaps.constants.BeatmapCountdown
import com.osudroid.beatmaps.constants.SampleBank
import kotlinx.coroutines.CoroutineScope

/**
 * A parser for parsing a beatmap's general section.
 */
object BeatmapGeneralParser : BeatmapKeyValueSectionParser() {
    override fun parse(beatmap: Beatmap, line: String, scope: CoroutineScope?) = splitProperty(line, scope)?.let {
        when (it.first) {
            "AudioFilename" -> beatmap.general.audioFilename = it.second
            "AudioLeadIn" -> beatmap.general.audioLeadIn = parseInt(it.second)
            "PreviewTime" -> beatmap.general.previewTime = beatmap.getOffsetTime(parseInt(it.second))
            "Countdown" -> beatmap.general.countdown = BeatmapCountdown.parse(it.second)
            "SampleSet" -> beatmap.general.sampleBank = SampleBank.parse(it.second)
            "SampleVolume" -> beatmap.general.sampleVolume = parseInt(it.second)
            "StackLeniency" -> beatmap.general.stackLeniency = parseFloat(it.second)
            "LetterboxInBreaks" -> beatmap.general.letterboxInBreaks = it.second == "1"
            "EpilepsyWarning" -> beatmap.general.epilepsyWarning = it.second == "1"
            "Mode" -> beatmap.general.mode = parseInt(it.second)
            "SamplesMatchPlaybackRate" -> beatmap.general.samplesMatchPlaybackRate = it.second == "1"
        }
    } ?: throw UnsupportedOperationException("Malformed general property: $line")
}
